File: 1.py
# Two-Sum returns the two indexes that sum up to a target, uses dictionary to do in O(n)!
##### Brute Force ######
# for i in range(len(nums)):
#     for j in range(i + 1, len(nums)):
#         if nums[i] + nums[j] == target:
#             return [i, j]
##### Brute Force ######


from typing import List


class Solution:
    # Subtracts number from the target and looks for it in dictionary,
    # if found okay, else add the number to dictionary and look for next one
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        lookup = {}
        for i in range(len(nums)):
            comp = target - nums[i]
            if comp in lookup:
                return [lookup[comp], i]
            else:
                lookup[nums[i]] = i

        return []


def main():
    arr = list(map(int, input("Input an array :").split()))
    target = int(input("Input Target :"))
    s1 = Solution()
    res = s1.twoSum(arr, target)
    print(res)


main()

File: 1.cpp
//Two sum problem, there are tow ways to do so.

// One is a nested loop first, that goes through each element
//looks for potential sums, if there is any returns those indeices, althoug not implemented here but sounds
//very intutive and simple.
// TC : O(N^2)
// SC: O(1)

#include<iostream>
#include<vector>
#include<unordered_map>
using namespace std;
class Solution {
public:

//The othere technique is below funciton which goes through vector, for every number, you check what will be the 
//other number required, by just doing "target - curr", now i now what is remaining, i will check in hash map, if 
//that very number is in map, if yes the so much so good, just print thath number, if not, puss this number in map
//as key and its indes as value and move on to next number, now as soon as you find the other number in map, it will 
//return that pair!!!
//TC :O(N)
//SC :O(N)
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mpp;
        
        for(int i=0; i<nums.size(); i++)
        {
            int curr = nums[i];
            int req = target - curr;
            
            if(mpp.find(req)!= mpp.end())
                return {i, mpp[req]};
            
            else
                mpp[curr] = i;
        }
        return {-1, -1};
    }
};

int main()
{
    Solution s1;
    vector<int> nums= {1, 2, 3, 4, 5, 6};
    vector<int> result = s1.twoSum(nums, 5);
    for(int i:result)
        cout<<i<<" "; 
}